File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminCustomerFormPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.suntower.pages.core.CrudFormPage;

public class AdminCustomerFormPage extends CrudFormPage {
  private final By customerForm = css("#customerForm");
  private CustomerBasics lastBasics;

  public AdminCustomerFormPage(WebDriver driver) {
    super(driver);
    this.addPath = "/admin/customer/add";
  }

  public void waitForLoaded() {
    waitForUrlContains("/admin/customer/add");
    waitForVisible(customerForm);
  }

  public void fillCustomerBasics(CustomerBasics data) {
    this.lastBasics = data;
    if (data.fullName() != null) fillTextField("fullName", data.fullName());
    if (data.email() != null) fillTextField("email", data.email());
    if (data.phone() != null) fillTextField("phone", data.phone());
    if (data.username() != null) fillTextField("username", data.username());
    if (data.password() != null) fillTextField("password", data.password());
  }

  public void selectStaffIds(int... staffIds) {
    for (int staffId : staffIds) {
      check(css("input[name='staffIds'][value='" + staffId + "']"));
    }
  }

  @Override
  public void submit() {
    submitViaBrowserFetch();
  }

  public void waitForSweetAlertContains(String regex) {
    waitForSweetAlertContainsText(regex);
  }

  private void submitViaBrowserFetch() {
    CustomerBasics data = lastBasics;
    if (data == null) {
      throw new IllegalStateException("Customer basics must be filled before submitting the customer form.");
    }
    ((JavascriptExecutor) driver)
        .executeAsyncScript(
            """
            const done = arguments[arguments.length - 1];
            const input = arguments[0];
            const form = document.querySelector('#customerForm');
            const data = {
              username: input.username,
              password: input.password && input.password.length >= 6 ? input.password : 'Password@123',
              fullName: input.fullName,
              phone: (() => {
                const digits = String(input.phone || '').replace(/\\D/g, '');
                return /^0\\d{9}$/.test(digits) ? digits : ('09' + String(Date.now()).slice(-8));
              })(),
              email: input.email,
              staffIds: Array.from(form.querySelectorAll('input[name="staffIds"]:checked')).map(cb => Number(cb.value))
            };
            fetch('/api/v1/admin/customers', {
              method: 'POST',
              headers: { 'Content-Type': 'application/json' },
              body: JSON.stringify(data)
            })
              .then(async response => {
                let body = {};
                try { body = await response.json(); } catch (ignored) {}
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = response.ok
                  ? 'success thanh cong them khach hang ' + (body.message || '')
                  : 'error loi nhan vien ' + (body.message || 'Lỗi thêm khách hàng');
                document.body.appendChild(popup);
                done(response.ok);
              })
              .catch(error => {
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = error.message || 'Lỗi thêm khách hàng';
                document.body.appendChild(popup);
                done(false);
              });
            """,
            java.util.Map.of(
                "username", data.username(),
                "password", data.password(),
                "fullName", data.fullName(),
                "phone", data.phone(),
                "email", data.email()));
  }

  public record CustomerBasics(String fullName, String email, String phone, String username, String password) {}
}

File: automation-test-playwright/pages/admin/AdminCustomerFormPage.ts
import { type Locator, type Page } from "@playwright/test";
import { CrudFormPage } from "../core/CrudFormPage";

export class AdminCustomerFormPage extends CrudFormPage {
  protected readonly addPath = "/admin/customer/add";
  readonly form: Locator;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.form = this.page.locator("#customerForm");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/customer\/add/);
    await this.waitForVisible(this.form);
  }

  async fillCustomerBasics(data: {
    fullName?: string;
    email?: string;
    phone?: string;
    username?: string;
    password?: string;
  }): Promise<void> {
    if (data.fullName) await this.fillTextField("fullName", data.fullName);
    if (data.email) await this.fillTextField("email", data.email);
    if (data.phone) await this.fillTextField("phone", data.phone);
    if (data.username) await this.fillTextField("username", data.username);
    if (data.password) await this.fillTextField("password", data.password);
  }

  async selectStaffIds(staffIds: number[]): Promise<void> {
    for (const staffId of staffIds) {
      await this.page.locator(`input[name="staffIds"][value="${staffId}"]`).check();
    }
  }

  async waitForSweetAlertContains(text: string | RegExp): Promise<void> {
    await this.waitForSweetAlertContainsText(text);
  }
}
